src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,125 @@
-
-            var projectDirectoryPath = TrimProject(projectFilePath);
-            if (string.IsNullOrEmpty(projectDirectoryPath))
-                return solutionFolders;
-
-            var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
-            if (string.IsNullOrEmpty(solutionFoldersPath))
-                return solutionFolders;
-
-            solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));
-
-            return solutionFolders;
         }
 
-        private IList<string> DetermineSolutionFolder(SlnFile slnFile, string fullProjectPath)
+        private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, string[] projectPaths, CancellationToken cancellationToken)
         {
-            if (_inRoot)
+            ISolutionSerializer serializer = SlnCommandParser.GetSolutionSerializer(solutionFileFullPath);
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
+            // set UTF8 BOM encoding for .sln
+            if (serializer is ISolutionSerializer<SlnV12SerializerSettings> v12Serializer)
             {
-                // The user requested all projects go to the root folder
-                return null;
+                solution.SerializerExtension = v12Serializer.CreateModelExtension(new()
+                {
+                    Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)
+                }); 
             }
-
-            if (_relativeRootSolutionFolders != null)
+            SolutionFolderModel solutionFolder = (!_inRoot && _solutionFolderPath != null)
+                ? solution.AddFolder(GetSolutionFolderPathWithForwardSlashes())
+                : null;
+            foreach (var projectPath in projectPaths)
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                // Get full project path
+                var relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
+                try
+                {
+                    AddProjectWithDefaultGuid(solution, relativePath, solutionFolder);
+                    Reporter.Output.WriteLine(CommonLocalizableStrings.ProjectAddedToTheSolution, relativePath);
+                }
+                catch (Exception ex)
+                {
+                    // TODO: Update with error codes from vs-solutionpersistence
+                    // Duplicate project
+                    if (Regex.Match(ex.Message, @"Project name '.*' already exists in the solution folder\.").Success || Regex.Match(ex.Message, @"Duplicate item '.*' of type 'Project'\.").Success)
+                    {
+                        Reporter.Output.WriteLine(
+                            string.Format(CommonLocalizableStrings.SolutionAlreadyContainsProject, solutionFileFullPath, relativePath));
+                        continue;
+                    }
```

Comment:
This is needed at the moment because the current behavior expects the command to succeed if a duplicate project is to be added, while vs-solutionpersistence throws an error.  This attempts to catch that error. 
In the future, there will be better ways of identifying these exceptions

Reply:
Oh didn't see that, will take a look :)